### PR TITLE
Port build system to Meson

### PR DESCRIPTION
**Pull request by eli-schwartz**
Prerequisite for porting pyerfa to Meson in https://github.com/liberfa/pyerfa/issues/91

Removing the autotools build can be done as well, if interested, implemented in eli-schwartz@0695ad204b5b58f06add71a4f908480bafb344c9 for an overall diffstat of:

```
 .gitignore        | 24 ------------------------
 Makefile.am       | 10 ----------
 RELEASE.rst       | 33 ++++++++++++++-------------------
 bootstrap.sh      |  8 --------
 configure.ac      | 32 --------------------------------
 erfa.pc.in        | 11 -----------
 m4/erfa-numver.m4 | 45 ---------------------------------------------
 meson.build       | 41 +++++++++++++++++++++++++++++++++++++++++
 src/Makefile.am   | 54 ------------------------------------------------------
 src/meson.build   | 51 +++++++++++++++++++++++++++++++++++++++++++++++++++
 10 files changed, 106 insertions(+), 203 deletions(-)
 delete mode 100644 .gitignore
 delete mode 100644 Makefile.am
 delete mode 100755 bootstrap.sh
 delete mode 100644 configure.ac
 delete mode 100644 erfa.pc.in
 delete mode 100644 m4/erfa-numver.m4
 create mode 100644 meson.build
 delete mode 100644 src/Makefile.am
 create mode 100644 src/meson.build
```

### CODE REVIEW

File: meson.build
Comment (mhvk) [inline meson.build:4]: Trying this on my laptop, I get warnings that the code uses features from a newer meson,
```
WARNING: Project targets '>=0.37.0' but uses feature introduced in '0.46.0': pkgconfig.generate optional positional argument.
Build targets in project: 3
WARNING: Project specifies a minimum meson_version '>=0.37.0' but uses features which were added in newer versions:
 * 0.46.0: {'pkgconfig.generate optional positional argument'}
```
Since we are introducing it here, I guess no reason not to pick a newer version? Maybe `>=0.55` to have `meson compile` available? In https://github.com/liberfa/pyerfa/issues/91#issuecomment-1256882425, a suggestion for 0.64 is made, but that is not yet out. My debian/testing has 0.63.2; debian/stable has 0.56.

File: RELEASE.rst
Comment (mhvk) [inline RELEASE.rst:80]: How about replacing this with `meson test -C builddir`? It calls the same thing, but can be used directly with `git bisect` according to the man page.

File: meson.build
Comment (mhvk) [inline meson.build:32]: Why not required? (Clearly do not understand yet how this works, so mostly out of curiosity!)

File: RELEASE.rst
Comment (mhvk) [inline RELEASE.rst:74]: Is it not still a good idea to clean the environment?

**Review by mhvk — COMMENTED** [review]
I had a bit of an "is this really needed?" feeling at first, but looking at the added code, the clarity of what is happening over the auto-config stuff seems like a major advantage!

Some comments in-line. A larger one is to add a new test block in `.github/workflows/ci_workflows.yml` to have this tested. A major advantage would seem to be that we can now include windows! (Which would remove the need for #75)

My sense would be to use this PR purely to *add* the meson capability and remove the auto-tools one in a follow-up PR. But maybe that is overkill. 

Let me ping @sergiopasra, @avalentino and @eteq, since probably good if they sign off on this.

File: meson.build
Comment (eli-schwartz) [reply meson.build:4]: This is because I set it to a low version specifically to check what warnings it generated, then after testing that this built okay I forgot to actually bump the version lol.

But of course since this is a demonstration and still needs to be hooked up to CI and get buy-in from stakeholders there's time for that. The most important thing to start with is to get something that works which people can look at.

File: RELEASE.rst
Comment (eli-schwartz) [reply RELEASE.rst:80]: The two commands are basically the same thing.
- `ninja test` first builds all artifacts, then executes `meson test --no-rebuild` to avoid recursion. It's designed to be a simple alias target that does what you probably want to do
- `meson test` will first try to run `ninja` to rebuild all targets, but can't rebuild when using the Visual Studio or xcode backends
  - `meson test` is a power user's frontend, and the main advantage of using it directly is that it has options to control test timeouts, wrap the tests with valgrind or gdb, select a subset of tests to run, etc.

Either one is fine, so we can use whichever one you prefer.

Actually for consistency with meson dist we might as well refer to meson test...

File: meson.build
Comment (mhvk) [reply meson.build:4]: Well, the demonstration is well on its way of convincing me at least! But hooking up to CI would be good.

File: meson.build
Comment (eli-schwartz) [reply meson.build:32]: For Windows support! :D

On Windows, there's no libm, and can't be. Whatever Windows support you have (I think there are math symbols already available? but I'm no Windows expert) will work without libm.

Libm is guaranteed to exist on POSIX platforms in some form, and to come with libc itself. In some cases it may actually be part of libc, in which case there may not be a file for it, but the compiler is required to accept an `-lm` argument which it may resolve to an empty static archive or simply delete from the command line, as it pleases -- POSIX doesn't mandate an implementation.

So, basically:
- libm may or may not exist as a library to look up
- if it is needed on that platform, it will always be there
- if it's not found, that is because it won't be needed

Comment (mhvk) [reply meson.build:32]: Ah, thanks!

File: RELEASE.rst
Comment (eli-schwartz) [reply RELEASE.rst:74]: Since Meson requires to build out of tree builds, built files from one build cannot contaminate another build. And meson has extremely accurate out of date detection, so you should never have to wipe your build directory and recreate it in order to trust that the correct version of the code was built.

Comment (mhvk) [reply RELEASE.rst:74]: OK, if this is basically guaranteed to be OK, then keeping it simpler sounds good to me!

**Comment by eli-schwartz** [comment]
FYI: I'll be offline for the holidays until Tuesday night, will push cleanups / CI changes after that.

**Comment by eli-schwartz** [comment]
Updated with additional CI workflows to test using meson instead of autotools.

File: .github/workflows/ci_workflows.yml
Comment (mhvk) [inline .github/workflows/ci_workflows.yml:53]: Shall we add windows too? Would be nice to show we can build on it!

File: .github/workflows/ci_workflows.yml
Comment (mhvk) [inline .github/workflows/ci_workflows.yml:71]: Maybe use the `meson` equivalent?

**Review by mhvk — COMMENTED** [review]
Looks good! A few small comments ...

File: .github/workflows/ci_workflows.yml
Comment (eli-schwartz) [reply .github/workflows/ci_workflows.yml:53]: I'm not overly familiar with Windows, but I took a stab at this.

My workflow test results are here: https://github.com/eli-schwartz/erfa/actions/runs/3148176374/jobs/5118436048

```
[255/255] Linking target src/t_erfa_c_extra.exe
FAILED: src/t_erfa_c_extra.exe src/t_erfa_c_extra.pdb 
"link"  /MACHINE:x64 /OUT:src/t_erfa_c_extra.exe src/t_erfa_c_extra.exe.p/t_erfa_c_extra.c.obj "/nologo" "/release" "/nologo" "/DEBUG" "/PDB:src\t_erfa_c_extra.pdb" "src\erfa.lib" "/SUBSYSTEM:CONSOLE" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib"
LINK : fatal error LNK1181: cannot open input file 'src\erfa.lib'
```

I think it's saying that because you don't have https://learn.microsoft.com/en-us/cpp/build/reference/module-definition-dot-def-files?view=msvc-170 it was unable to build a shared DLL that contains public symbols (MSVC hides all symbols by default IIRC, unlike unix compilers that expose them all by default), although configuring meson with `-Ddefault_library=static` would produce a static-only library that would likely work.

Comment (eli-schwartz) [reply .github/workflows/ci_workflows.yml:49]: Okay, pushed that with hardcoding the static library for now on Windows.

File: meson.build
Comment (eli-schwartz) [reply meson.build:32]: CI results for Windows are e.g. here: https://github.com/eli-schwartz/erfa/actions/runs/3148215060/jobs/5118511545

```
C compiler for the host machine: cl (msvc 19.33.31630 "Microsoft (R) C/C++ Optimizing Compiler Version 19.33.31630 for x64")
C linker for the host machine: link link 14.33.31630.0
Host machine cpu family: x86_64
Host machine cpu: x86_64
Library m found: NO
```

But in the end it fully compiles and runs the selfcheck executables successfully, without an `m` library. The functions needed by erfa are part of the CRT.

File: .github/workflows/ci_workflows.yml
Comment (mhvk) [reply .github/workflows/ci_workflows.yml:67]: Obviously, it really doesn't matter though I find it easier to just use the one tool and let it do what it wants under the hood, so, `meson compile -C builddir` -- this is what the documentation suggests too: https://mesonbuild.com/Running-Meson.html

Comment (eli-schwartz) [reply .github/workflows/ci_workflows.yml:67]: The documentation simply states both are possible.

`meson compile` exists basically for one reason: if you don't know or care what build backend you used, it will detect whether to run ninja or msbuild or xcodebuild for you.

To do that, it imports 70k lines of code with the only active component being `subprocess.run([detect_ninja(), '-c', build_dir])`

Comment (mhvk) [reply .github/workflows/ci_workflows.yml:67]: OK, happy to stay with what you have!

File: .github/workflows/ci_workflows.yml
Comment (mhvk) [inline .github/workflows/ci_workflows.yml:5]: Need to remove this before we merge!

**Review by mhvk — COMMENTED** [review]
I think this now looks great -- this is a very nice improvement! Remains to revert the edits to when CI runs.

Also, as I'm not very familiar with the build system and this changes it, I'd like @eteq and @sergiopasra to have a look as well.

File: .github/workflows/ci_workflows.yml
Comment (eli-schwartz) [reply .github/workflows/ci_workflows.yml]: Done. It's a trivial rebase because I had it as a distinct commit. Actually I thought I had already done so.

**Review by mhvk — APPROVED** [review]
That was quick, thanks! Approving now but waiting a little while with merging to give the others a final chance to have a look.

**Comment by sergiopasra** [comment]
It looks good to me. I would like to suggest to keep autotools machinery for one major release (what means update both meson and autotools files when releasing) and the remove the autotools files.

**Review by sergiopasra — APPROVED** [review]

**Comment by eli-schwartz** [comment]
> I would like to suggest to keep autotools machinery for one major release

Yep, that's generally a good idea when migrating build systems because you don't want to leave people in the lurch if for some reason things don't work for them. If anyone reports bugs, they can be fixed before dropping the old build system.

... granted, this build isn't that complex, so one could argue the risks are low. But I didn't want to impose, hence why I specifically left that out of this PR. :)

**Comment by eli-schwartz** [comment]
Sorry, I realized that I neglected to update README.rst as well.

**Comment by mhvk** [comment]
In my mind, this was already merged, but clearly I forgot!  So, let's get it in now. Thanks so much, @eli-schwartz!

**Comment by eli-schwartz** [comment]
Great, happy to help. I guess this means I can now look at pyerfa for you. ;)

**Pull request merged**
